src/interpreter/executors/main.rs

```diff
@@ -27,171 +45,40 @@ impl Interpreter<PredicateStorage> {
     ///
     /// This is not a valid entrypoint for debug calls. It will only return a `bool`, and not the
     /// VM state required to trace the execution steps.
-    pub fn check_predicates(tx: CheckedTransaction, params: ConsensusParameters) -> bool {
+    pub fn check_predicates<S, Tx>(checked: Checked<Tx, S>, params: ConsensusParameters) -> bool
+    where
+        S: Stage,
```

Comment:
The VM doesn't really care about the stage, since any level of checked must be guaranteed to be usable by the VM. Is there a way we could avoid the viral spread of this generic across the VM which doesn't need to distinguish between txs with checked vs unchecked signatures?

Reply:
Maybe the VM could have it's own simplified checked type and these methods could accept any `Tx: Into<VmCheckedTx>`?

Reply:
The first question is, do we need `Stage` in the type of `Checked` at all? I introduced `Stage` because I thought that:
- VM works only with fully checked transactions.
- TxPool works with partially checked transactions.
- During block production, we transform all transactions from TxPool into fully checked and pass them into VM.
- In the future, we will have a separate stage for verified predicates.
- In the future, we can have a stage for checked not finalized transactions(during the production of the block, we fill several fields like placeholders for the contract's utxo, outputs, etc).

But after a discussion with you, VM can work with any type of checked transaction. TxPool wants to work with fully checked transactions(for now, it works with both types). We don't want to make the code more complex.

The idea was to have a state machine for transaction checking on rust level type to manage our code in the right direction. And allow only a limited set of methods for each stage.

So, we can avoid state machine and instead use an enum for `Checked` to only highlight different stages.

```rust
pub enum Checked<Tx: IntoChecked> {
    Stateless(Tx, Tx::Metadata),
    Signatures(Tx, Tx::Metadata),
    ..,
}
```

The problem with this approach is that some methods will do nothing, like the `check_signatures` for the `Signature` variant.

Also, we can use the old approach and have only one struct `Checked<Tx>` with booleans inside or bitmask to describe checking stages.

Reply:
> The VM doesn't really care about the stage, since any level of checked must be guaranteed to be usable by the VM. Is there a way we could avoid the viral spread of this generic across the VM which doesn't need to distinguish between txs with checked vs unchecked signatures?

It affects only input methods like `deploy`, `transaction`, and `check_predicates`. So the spread is not huge=)

> Maybe the VM could have it's own simplified checked type and these methods could accept any Tx: Into<VmCheckedTx>?

We can create a `CheckedForVM` or `VmChecked` type, but it will spread usage of `into()` during the call of methods above=) And it seems worse the definition of 3 methods with generic

Reply:
If we don't need stage at all then maybe we should just remove it to reduce complexity. It's probably not a big deal if check_signatures does nothing for a `Checked::Signatures` variant, as that could be optimized away by the compiler potentially.

Reply:
From the performance side, of course, there is no problem, it will be optimized, and the call will be removed at all. It is more about code clarity=)

Then let's remove it for now and return back to this idea later

Reply:
Done https://github.com/FuelLabs/fuel-tx/pull/197/commits/c9b86b8a6a0ce1bfa123afffcd6b07a3d866519f